crates/circuit/src/dag_circuit.rs

```diff
@@ -6223,6 +6223,16 @@ impl DAGCircuit {
         self.cargs_interner.get(index)
     }
 
+    /// Insert qargs, return the intern index
+    pub fn set_qargs(&mut self, qargs: &[Qubit]) -> Interned<[Qubit]> {
+        self.qargs_interner.insert(qargs)
+    }
+
+    /// Insert cargs, return the intern index
+    pub fn set_cargs(&mut self, cargs: &[Clbit]) -> Interned<[Clbit]> {
+        self.cargs_interner.insert(cargs)
+    }
```

Comment:
IMO, we should favor adding a variant of `push_back` that deals in terms of native bits (in addition to the one that takes a `PackedInstruction`) rather than exposing methods like these.

(I can't think of much use for mutable interners other than the case of emplacing a new instruction into the DAG, and other passes will likely need this functionality too anyway 🙂.)

Reply:
To make sure I fully understand the suggestion: you are proposing to add a function like
```
fn push_back_op(
    op: &PackedOperation, 
    qargs: &[Qubit], 
    cargs: &[Clbit], 
    params: &Option<Box<SmallVec<[Param; 3]>>>, 
    extra_attrs: &Option<Box<ExtraInstructionAttributes>>, 
    #[cfg(feature = "cache_pygates")] 
    py_op: OnceCell<Py<PyAny>>
)
```
Internally the function would do the interning, create a packed instruction, and call the existing `push_back`. Is this on the right track? (Btw, I did not yet try to implement it, but the "optional" `py_op` already scares me).

Though in my particular use-case, I would like to avoid de-interning and re-interning things related to classical bits, since the clbits iterner is copied in `copy_empty_like` and the classical bits are not remapped by the pass. What do you think?

Reply:
I think the signature for this should be something like:

```rust
fn apply_operation_back(
    op: &PackedOperation, 
    qargs: &[Qubit], 
    cargs: &[Clbit], 
    params: Option<&[Param]>,
    extra_attrs: Option<&ExtraInstructionAttributes>,
) -> NodeIndex
```

Note that I used the name `apply_operation_back` (I called the Python version `py_apply_operation_back` with anticipation of this Rust-oriented version).

I think the `py_op` cache will get updated on demand within the instruction, so perhaps we can get away with keeping it out of the interface here. Maybe @jakelishman has thoughts?

Reply:
Issue #13134 has been create for this work. I will quote the above discussion there

Reply:
I'm fine with the method as described by Kevin, but I also think there ought to be ways to push operations back while passing an already-known `Interned` key.  As in Sasha's case, there's plenty of cases where we need to pass stuff in where we can know that the interned key is safe, and looking up and then re-hashing the `&[Clbit]` or `&[Qubit]` would be a waste of cycles.

_Personally_ I have no problem with exposing a method to allow manual insertions into the interners as well - I don't think it should be the standard way to do that, but you can't break data coherence by doing it, since interners are append-only, and adding the key doesn't invalidate existing keys.  That said, I don't immediately have uses for that in mind - I have a rough imagined optimisation to the Sabre post-routing rebuilder that pre-allocates the interner, but no proof it'd be a good idea - so I'm not pushing for adding them.

Reply:
Just a heads up @alexanderivrii , #13143 is now merged which adds a native `apply_operation_back`. 

> Though in my particular use-case, I would like to avoid de-interning and re-interning things related to classical bits, since the clbits iterner is copied in copy_empty_like and the classical bits are not remapped by the pass. What do you think?

Given that you have this requirement, I'd say feel free to use `DAGCircuit::push_back` directly instead, but please also use the interner directly (e.g. `dag.qargs_interner().insert(mapped_qargs)`) rather than adding these `set_{q|c}args` methods to `DAGCircuit`.

Reply:
_“If you wait by the river long enough, the bodies of your enemies will float by.”_
― Sun Tzu

@kevinhartman, @raynelfss, and @jakelishman -- many thanks for the insightful discussion and adding this feature.

Reply:
> does not seem to work due to mutability concerns

Ah, you're right. I forgot that we don't expose a mutable reference for those. We _could_, but if you're happy to just use `apply_operation_back` let's leave that up to a future decision.

> First, how should we pass params to apply_operation_back: as simply cloning it leads to the error

I believe what you're looking for is something like: `params.as_deref().cloned()`. This takes you from an `Option<Box<SmallVec<[Param; 3]>>>` to `Option<&SmallVec<[Param; 3]>>` and finally to an owned `Option<SmallVec<[Param; 3]>>`.

For `py_op`, you can clone whatever is in the existing `PackedInstruction` (though you will have to do it conditionally with the macro, just like how the parameter itself is declared in `DAGCircuit::apply_operation_back`).

In general, the `py_op` thing is just a cached version of the Python-based operation, so if you have a `PyObject` corresponding to the `PackedOperation` you're pushing, you can pass it. Else, just pass `None`.

Reply:
Thanks, this is done. Now the only change to the dag circuit interface in this PR is making `copy_empty_like` public.